### ISSUE TITLE
Combining AxisArrays - beyond cat

### ISSUE THREAD

**Issue by GordStephen** [mbauman/AxisArrays.jl#27]
Thinking more about high dimensional concatenation and some typical use cases for combining AxisArrays, there may be a few ways to make better use of axis data for "smarter" alternatives to `cat`:

`merge{T<:AxisArray}(As::T...)`: given AxisArrays with identical axes but disjoint / non-colliding axis values, automatically combine the inputs into a single AxisArray spanning at least the discrete ranges of all of the input AxisArrays. Unlike a basic `cat`, dimensional axes would be automatically sorted as needed. Any elements at a newly-created axis coordinate would assume some default value (in Base a comparable multidimensional `cat` fills undefined elements with zeros, but maybe this value could be user-definable?).

`join{T<:AxisArray}(method::Symbol, As:T...)`: given AxisArrays with identical axes and possibly overlapping axis values, automatically concatenate the inputs on a new, higher dimension/axis. Dimensional axes would be sorted appropriately. Inner/left/right/outer join methods would determine whether all elements are included in the returned AxisArray or just those at axis values found in all/first/last input AxisArrays.

Thoughts? Are these actions generally common enough to merit special functions? Like the `cat` work I was going to implement something similar for a downstream package but figure it may be more generally useful?

**Comment by mbauman** [comment]
I'm definitely happy to see more functionality added here, in general.  Specifically, I think it'd be amazing to bring AxisArrays and DataFrames closer together with a shared or analogous interface.  These functions are definitely a part of getting there.

For `merge` — I'm having a hard time imagining how there'd be missing slots.  I think I'd extend `Base.merge` and follow the description there, using the axis values as "keys".

**Comment by GordStephen** [comment]
`merge` would essentially imply an outer join on all existing dimensions, so for cases like

```julia
julia> A = AxisArray(randn(5,1), Axis{:Time}([1,2,3,4,5]), Axis{:Category}([:A]))
2-dimensional AxisArray{Float64,2,...} with axes:
    :Time, [1,2,3,4,5]
    :Category, [:A]
And data, a 5x1 Array{Float64,2}:
 -1.0438  
  1.27203 
 -0.252605
 -0.814517
 -0.428174

julia> B = AxisArray(randn(5,1), Axis{:Time}([3,4,5,6,7]), Axis{:Category}([:B]))
2-dimensional AxisArray{Float64,2,...} with axes:
    :Time, [3,4,5,6,7]
    :Category, [:B]
And data, a 5x1 Array{Float64,2}:
 -0.96318 
 -0.358768
 -0.825307
  0.22487 
 -1.44375 
```

values in the output of `merge(A, B)` at (1, :B), (2, :B), (6, :A), and (7, :A) would be undefined.

**Comment by mbauman** [comment]
Ah, that makes sense.  I suppose there are actually quite a few options in that case:

* Punt to Base's uninitialized default — this might become zeros (or at least deterministic) some day, but for now it'll be gibberish.
* Use zeros — this will fail for some eltypes.
* Return a wrapped NullableArray if any axes don't match — this is a little heavy-handed, but could actually be done in a type-stable way with a generated function.  It'll probably be pretty surprising, though.
* Always return a wrapped NullableArray.  I've not done any experiments with AxisArrays{NullableArray}, but a convenience function could be provided to pop the NullableArray and go back to an AxisArray{Array} if there aren't any null values.
* Error — this is really heavy-handed, but it gives us the most leeway for changes in the future.

I think I'd lean towards the first or last option, but I'll defer to your preference here since you're actually the one working towards dog-fooding this. :)

**Comment by GordStephen** [comment]
How about a keyword argument that defaults to zero?
```julia
merge{T}(As::AxisArray{T}...; fillvalue=zero(T))
```

**Comment by mbauman** [comment]
Sure, that works.  Thanks again for the thoughtful contributions!

**Comment by GordStephen** [comment]
Merged #28
